docs: add a how-to for setting open ports (#1579)
canonical/operator: docs/howto/manage-opened-ports.md

(manage-opened-ports)=
# How to manage opened ports
> See first: {external+juju:ref}`Juju | Hook Commands | open-port <hook-command-open-port>`

> Juju manages the IP of each unit, so you need to instruct Juju when you need to
> have a stable address on which to reach the charm. Typically, charms manage this
> by offering to integrate with an ingress charm, but you may also wish to have
> the charm itself open a port.

Reviewer: I'd probably shorten this (mainly to avoid two "need to"s):

```suggestion
Juju manages the IP of each unit, so you need to instruct Juju
if you want the charm to have a stable address. Typically, charms manage this
by offering to integrate with an ingress charm, but you may also wish to have
the charm itself open a port.
```

## Implement the feature

> `ops` offers [](ops.Unit.open_port) and [](ops.Unit.close_port) methods, but
> typically it will be simpler for your charm to declare which ports should be
> open. For example, to set an open TCP port based on a configuration value, in
> your `src/charm.py` file, in your `config-changed` observer:

Reviewer: How about mentioning open_port and close_port after the example instead? I feel it's clearer to directly start with the part that we'd prefer people to use - something like this:

```suggestion
It's simplest for your charm to use [](ops.Unit.set_ports) to to declare which ports
should be open. For example, to set an open TCP port based on a configuration value,
do the following in your `config-changed` observer in `src/charm.py`:
```

Then the link to set_ports after the code could be replaced by a mention of open_port and close_port as alternatives.

```python
def _on_holistic_handler(self, _: ops.EventBase):

>     port = cast(int, self.config['server-port'])

Author: Come along to my "validating data" chat next week and get back to me after that :)

>     port = cast(int, self.config['server-port'])

Reviewer 2: Looking forward to it! Didn't realise you were tacking config as well.

    self.unit.set_ports(port)
```

> See more: [](ops.Unit.set_ports)

> Examples: [mysql-k8s opens the MySQL ports](https://github.com/canonical/mysql-k8s-operator/blob/a68147d0fbf66386ab087f4cfcc19784fcc2be6e/src/charm.py#L648), [tempo-coordinator-k8s opens both server and receiver ports](https://github.com/canonical/tempo-coordinator-k8s-operator/blob/ece268eae1158760513807a02972c138fd39afcf/src/charm.py#L95)

## Test the feature

> > See first: {ref}`get-started-with-charm-testing`

Reviewer: I recommend we remove this link from this point, and instead put it in the "See first" blocks for unit tests and integration tests (both sections). For a couple of reasons:

- Fewer stops and starts when reading through the page.
- If someone scans the TOC and jumps straight to the section on integration testing, they'll still see the link to charm testing.

> You'll want to add unit and integration tests:

Reviewer: I think a period is fine here.

```suggestion
You'll want to add unit and integration tests.
```

### Write unit tests

> See first: {ref}`write-scenario-tests-for-a-charm`

> In your unit tests, include any ports that are already open when the event is
> run in the [](ops.testing.State.opened_ports) component of the input `State`
> (any ports not listed are assumed to be closed). After events that modify which
> ports are open, assert that the output `State` has the correct set of ports. 

Reviewer: I found the first sentence slightly confusing. Would it be correct to change it to the following?

```suggestion
In your unit tests, use the [](ops.testing.State.opened_ports) component of the
input `State` to specify which ports are already open when the event is
run. Ports that are not listed are assumed to be closed. After events that modify which
ports are open, assert that the output `State` has the correct set of ports. 
```

For example, in `tests/unit/test_charm.py`, this verifies that when the
`config-changed` event runs, the only opened port is 8000 (for TCP):

```python
def test_open_port():
    ctx = testing.Context(MyCharm)
    state_in = testing.State()
    state_out = ctx.run(ctx.on.config_changed(), state_in)
    assert state_out.opened_ports == {testing.TCPPort(8000)}
```

### Write integration tests

> See first: {ref}`write-integration-tests-for-a-charm`

To verify that the correct ports are open in an integration test, deploy your

> charm as usual, and then try to connect to the appropriate ports.

Reviewer: Nice :smile: In this sense, integration tests are so much simpler...!

> Added to your `tests/integration/test_charm.py` file, this test verifies a charm
> that opens a port specified in the configuration, but prohibits using port 22:

Reviewer: Suggest a minor rephrasing:

```suggestion
By adding the following test to your `tests/integration/test_charm.py` file, you can verify
that your charm opens a port specified in the configuration, but prohibits using port 22:
```
